### ISSUE TITLE
Sort Expressions Using Query Aliases

### ISSUE THREAD

**Issue by cmoesel** [cqframework/clinical_quality_language#127]
I’m chasing down an issue in the coffeescript execution engine, but it’s led me to question some things regarding sort.

# The Problem

In the R1.1 spec, we introduce sort with this example:
```
[Encounter: “Inpatient”] E sort by start of period
```
I think this is the way sort by is intended to be used – directly referencing a property in each of the returned items.

That said, we also have some examples further on that do things like this:
```
First([Condition: "Acute Pharyngitis"] C sort by C.onsetDateTime desc)
```
Note the use of the `C` alias in the sort expression.  I’m not sure this is actually valid and it raises the question: **Can a query alias be referenced in a sort clause?**

At first glance, this seems OK – but when you really think about it, `C` is an alias used during the query execution.  The sort is defined to operate on the _results_ of the query – _after_ the execution has occurred – at which point `C` doesn’t really have meaning.

Looking at the CQL 1.1 spec, section 5.3.3 states:
> After the iterative clauses are executed for each element of the query source, the sort clause, if present, specifies a sort order for the final output.  This step simply involves sorting the output of the iterative steps…

Since the alias only holds meaning during the iterative clauses (and changes during each iteration), then if sort comes _after_ iteration, the alias has no meaning.  Section 5.3.4 makes this ordering even more clear when it lists out the steps to implementing a query: ForEach, Times, Filter, Distinct, Sort.  It seems that by the time you get to _Sort_, the alias shouldn't be in scope anymore.

Another reason it likely doesn’t make sense to reference an alias in the sort clause is that it introduces a possibility for ambiguity when the thing you’re sorting has a property with the same name as your alias.  Consider this:
```
({Tuple{E: 1}, Tuple{E: 3}, Tuple{E: 2}}) N sort by E
```
Like the first example above, this shows how sort is intended to work -- by referencing a _property_ in the returned items.  This is fine, but what happens when we do this instead:
```
({Tuple{E: 1}, Tuple{E: 3}, Tuple{E: 2}}) E sort by E
```
If we allow aliases in the sort expression, what does the `E` in the sort mean?  Is it the _property_ `E` or is it the item `E`?  It’s unclear.

# Where It's Currently Allowed

Despite the above, there are a few things that hint that using aliases in a sort clause is allowed:
* the spec has a number of examples where an alias is used in a sort clause
* the cql-to-elm translator _allows_ aliases in sort clauses (they become `AliasRef` instances)

If it's not really allowed, these will have to be fixed.

# A Further Issue / Limitation

That said, there are a few cases where we _do_ want to reference the returned _item_ in the sort clause rather than a _property_ of the item.  The use case that led me here is that I want to sort items by the result of a function that takes the item as a parameter.  A trivial example is:
```
({1,-3,2,-5,4}) N sort by Abs(N)
```
Now... if referencing an alias is not allowed, then what do I pass into the `Abs` function?  It seems we would need to have some sort of inferred variable, such as:
```
({1,-3,2,-5,4}) N sort by Abs($result)
```
Right now, I can use a workaround, but it is _ugly_:
```
(({1,-3,2,-5,4}) N return Tuple{num: N} sort by Abs(num)) N2 return N2.num
```

# So...

I guess that leaves the following questions:
* Should aliases be allowed in sort expressions?
   * If yes...
      * How do we reconcile that with the implementation guidance in the spec?
      * How do we specify the expected behavior in the ambiguous example I provided?
   * If no...
      * How do we support sort expressions that want to operate directly on the _result_?

**Comment by brynrhodes** [comment]
You're right that this is somewhat ambiguous in the spec and should be clarified. Given that one of the primary targets of ELM is simplicity of implementation, we should be consistent with the implementation guidance and _not_ allow alias references in the sort. I checked the translator and it does incorrectly allow alias references, but that would be a challenge when it came to implementation because the source data isn't even available at that point in the query pipeline.

So, I propose that we use this issue to fix the translator so that it no longer allows references to aliases (which based on the implementation guidance is the intended interpretation for CQL), and that we submit STU comments against 1.2 to correct the examples that allow alias references as errors in the specification.

**Comment by cmoesel** [comment]
That sounds good to me.  Any thoughts on how we would support referencing the returned item _as a whole_ in the sort clause?

My real-world use case is that when I sort FHIR `Observation`s by date, I want to sort by the "best" date I have for each -- so it's essentially a sort by `Coalesce(effectiveDateTime.value, effectivePeriod."end".value, effectivePeriod."start".value, issued.value)`.  This _works_, but there is other logic where I want to use this same _Coalesce_ expression on Observations, so I put it into a `FindObservationDate` function that takes the `Observation` as an argument.  Ideally I'd use `FindObservationDate` in my sort clause, but I can't since the argument is the full Observation item. 

Make sense?

**Comment by brynrhodes** [comment]
Yes, and sorry, I meant to address that in my response as well but forgot. I propose we use the nominated iteration variable $this that was introduced in 1.2. So within a sort clause expression, you'd be able to access the current item using $this.

**Comment by cmoesel** [comment]
Merged in the fix with #129

**Comment by brynrhodes** [comment]
Submitted comment [1436](http://www.hl7.org/dstucomments/showdetail_comment.cfm?commentid=1436)
